### ISSUE TITLE
ridiculous idea: 'breakpoints'

### ISSUE THREAD

**Issue by ladyada** [adafruit/circuitpython#298]
this doesnt exist in python i think but its an interesting idea - can we pop out from the main.py code to REPL but still have access to that context?

**Comment by tannewt** [comment]
We could. Micropython actually worked that way originally where the heap
was shared. There was no way to continue main though. Would the goal be for
debugging?

**Comment by ladyada** [comment]
yeah like "stop here i want to see my variable state" - im ok with not being able to re-continue

**Comment by caternuson** [comment]
> this doesnt exist in python i think
try `python -i foo.py`

**Comment by dhalbert** [comment]
A PEP (Python Enhancement Proposal) for a builtin function called `breakpoint()`: https://www.python.org/dev/peps/pep-0553/. Will probably be in Python 3.7.

**Comment by dhalbert** [comment]
yeah - it provides us a syntax that will be the standard

**Comment by tannewt** [comment]
We could also reach out upstream to see if its development we could fund.

**Comment by dastels** [comment]
I've started looking into this. Initially I'm playing in the unix port, but the code should all live in py/

https://github.com/dastels/circuitpython

**Comment by ladyada** [comment]
rad. to extend my terrible idea...would it ever make sense to have a debugging backchannel thru HID...so REPL stays clean

**Comment by dastels** [comment]
I think then you start thinking about watchpoints, tracing, single stepping, and the like. And that will need far more invasive changes. Certainly doable, but might not be in line with a simple, educational environment.  

My goal is to be able to drop in a debugger() or breakpoint() (Smalltalk used halt() ) function call and let you poke around the local environment, possibly peek into different frames on the stack.  I'd also like to support passing an boolean expression to debugger() (or whatever) to allow conditional breakpoints. That should be trivial to add once the basics are in place.

**Comment by ladyada** [comment]
yah agree- i was just thinking to the far future :) for now breakpoints are enough!

**Comment by dastels** [comment]
I'll probably explicitly NOT support using `def` in the debug repl. Expressions, yes. Assignment to in-scope variables, yes with a warning that it might mess things up. I'd like to support looking at variables in prior stack frames. I'm not sure how that would look yet... but that's not for a while.  The next step is just to get local variable (and parameter) references working.

**Comment by tdicola** [comment]
Would re-enabling stopping of main and keeping the context be a simple interim fix in 3.0 or earler?  It's handy to debug your main.py by Ctrl-C and looking at variable values.

**Comment by dastels** [comment]
My first thought it that where you stop it would be random. Even in a simple loop where you stop would impact what you expect values to be.

**Comment by ScottDWebster** [comment]
It would be nice if there was something like DBGP (https://xdebug.org/docs-dbgp.php) to allow remote debugging.

**Comment by hybotics** [comment]
Hi,

I am VERY interested in seeing some kind of source level debugging added to Circuitpython. By this I mean the facility to have a script stop at a certain point (breakpoint), be able to examine and change variables, and the continue after the breakpoint. I am at a point with one script where it would be VERY useful to be able to do this.

For instance, I have verified that code sequences work, by using the REPL, but I am not able to see what my script is actually doing and why it is not working right. This is where I think source level debugging would come in VERY handy. Print statements just are not giving me what I need.

8-Dale (aka geekguy on the forums)

**Comment by kevinjwalters** [comment]
I did a search on forums and found [Adafruit Forums: Is using circuitpython on rPi the only way to use a debugger](https://forums.adafruit.com/viewtopic.php?f=60&t=155219).

I have some moderately complex code now in CircuitPython and this has some bugs in it. Normally this would be when I would add a breakpoint and then single-step the (python) code and I think I'd be able to work it out in a few minutes...

I think this is going to become more common and more of a gap as the bigger CPUs like SAMD51 and nRF52840 allow more complex programs to be written.

**Comment by canyon289** [comment]
Support for an interactive repl debugger like `breakpoint()` or `import pdb; pdb.set_trace()' would be nice, but I can get away with print statements in my case for now :)

**Comment by calcut** [comment]
Just adding my interest in this feature.

**Comment by ladyada** [comment]
https://www.youtube.com/watch?v=RRN_ngx01vM
